HIVE-25006: Commit Iceberg writes in HiveMetaHook instead of TezAM

marton-bod opened the pull request:
Note: CI build will fail since Tez dependency is SNAPSHOT, using: https://github.com/apache/tez/pull/101

- `TezTask` queries the DAGClient to get each reducer (or mapper, if map-only) vertex's ID and number of tasks. There's some necessary translation between vertex ID and job ID.
- It puts this info into the session conf suffixed by the table name (for multitable insert cases). For non-Iceberg table writes, this info should not be populated.
- This jobId/taskNum info is retrievable from the MetaHook side, which can thus create the job context and commit the operation.

Assumption: for each table, there is only one vertex that writes to it.

marton-bod commented:
@pvary @lcspinter can you please take a look?

pvary commented on iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:
```diff
@@ -250,4 +259,43 @@ private static PartitionSpec spec(Schema schema, Properties properties,
+      throws MetaException {
+    // construct the job context
+    JobConf jobConf = new JobConf(conf);
+    String tableName = TableIdentifier.of(table.getDbName(), table.getTableName()).toString();
```
How this will get the `Catalog` information?

marton-bod replied:
Not sure yet, I'll look into it. The catalog name might be stored in the table params, or we might need to load the Iceberg table to find out if we can. @lcspinter might have some more insights

pvary commented on iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:
```diff
@@ -250,4 +259,43 @@ private static PartitionSpec spec(Schema schema, Properties properties,
   }
+
+  @Override
+  public void commitInsertTable(org.apache.hadoop.hive.metastore.api.Table table, boolean overwrite)
```
In case of multi-table insert this method is called for every table?

marton-bod replied:
yes

pvary commented on iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:
```diff
@@ -121,6 +121,9 @@ public void configureJobConf(TableDesc tableDesc, JobConf jobConf) {
       jobConf.set("mapred.output.committer.class", HiveIcebergOutputCommitter.class.getName());
+      // this will turn off job committing on the Tez AM side, so that we can commit the write jobs
+      // on HS2 side instead using the HiveIcebergMetaHook
+      jobConf.set("hive.tez.mapreduce.output.committer.class", "");
```
nit: is there a constant for the key? Also I have seen `NullOutCommitter` class around somewhere. Could we use that here?

marton-bod replied:
Good idea, will try the NullOutCommitter

marton-bod replied:
I replaced the string key with the constant. As for using a null committer, that class would need to extend the `OutputCommitter` from the `tez-api` package. Currently iceberg-handler does not depend on Tez, so we can only achieve that if we introduce that dependency.

pvary replied:
Then using `NullOutCommitter` does not worth it

pvary commented on iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:
```diff
@@ -291,8 +290,6 @@ public void testSelectDistinctFromTable() throws IOException {
 
   @Test
   public void testInsert() throws IOException {
-    Assume.assumeTrue("Tez write is not implemented yet", executionEngine.equals("mr"));
```
Ohh.. how I like these changes 😄 🥇

marton-bod commented on iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java:
```diff
@@ -224,6 +224,11 @@ private void cleanup(JobContext jobContext) throws IOException {
     // If there are reducers, then every reducer will generate a result file.
     // If this is a map only task, then every mapper will generate a result file.
     int expectedFiles = conf.getNumReduceTasks() > 0 ? conf.getNumReduceTasks() : conf.getNumMapTasks();
+    // For Tez, we can only reliably get the number of tasks from the Tez AM, not the job conf
```
I wanted to be considerate of MR (even though it's unsupported in upstream Hive), but my initial approach was to not touch it here but set the numReduceTasks to the correct value in the `commitInsertTable`

pvary replied:
If I understand correctly in Hive3 `commitInsertTable` is not called when the execution engine is MR.
If it would be called anywhere then we are in a big trouble anyway as `HiveIcebergOutputCommitter.commitJob` would be called twice.

So all-in-all I would go for this change

marton-bod replied:
Sounds good. Will set the value in the `commitInsertTable`

pvary commented on iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveShell.java:
```diff
@@ -154,6 +155,13 @@ public void closeSession() {
     }
   }
 
+  public String executeExplain(String statement) {
+    List<Object[]> objects = executeStatement(statement);
+    return objects.stream()
+        .map(o -> (String) o[0])
+        .collect(Collectors.joining("\n"));
+  }
+
```
👍

marton-bod replied:
Needed this during way too many debug sessions :)

pvary replied:
Leave a comment there not to remove it even if it is not used 😄

pvary commented on ql/src/java/org/apache/hadoop/hive/ql/exec/tez/TezTask.java:
```diff
@@ -250,9 +255,32 @@ public int execute() {
+          boolean hasReducer = allWork.stream().map(workToVertex::get).anyMatch(v -> v.getName().startsWith("Reducer"));
+          for (BaseWork baseWork : allWork) {
+            Vertex vertex = workToVertex.get(baseWork);
+            if (!hasReducer || vertex.getName().startsWith("Reducer")) {
```
Do I understand correctly that we would like to identify if we wrote something or not with this vertex?
Would `Vertex.getDataSinks()` work here?

What about if we have a multi-table insert writing to Iceberg and non-Iceberg tables?

marton-bod replied:
Correct, the goal is to pick out only those vertices that wrote to a table. I will look into the `dataSinks()` method, that could be a better way to identify them. 

In principle, the mixed table scenario you mentioned should work. If we had two reducer vertices (that wrote to tables), we'll check both of them here. The vertex that wrote to the Iceberg table will publish its info (jobId, task num) with the table suffix into the session conf, and the vertex that wrote the Hive table will do the same except without any table suffix. But since we need no commit for the Hive table one, that second part is more or less irrelevant for now.

pvary commented on ql/src/java/org/apache/hadoop/hive/ql/exec/tez/TezTask.java:
```diff
@@ -250,9 +255,32 @@ public int execute() {
+              // status.getId() returns something like: vertex_1617722404520_0001_1_00
+              // this should be transformed to a parsable JobID: job_16177224045200_0001
+              int vertexId = Integer.parseInt(jobIdParts[jobIdParts.length - 1]);
+              String jobId = String.format("job_%s%d_%s", jobIdParts[1], vertexId, jobIdParts[2]);
```
nit: Maybe use a constant here

marton-bod replied:
Done

pvary commented on ql/src/java/org/apache/hadoop/hive/ql/exec/tez/TezTask.java:
```diff
@@ -250,9 +255,32 @@ public int execute() {
+              String jobIdKey = HIVE_TEZ_COMMIT_JOB_ID + (tableName == null ? "" : "." + tableName);;
+              String taskCountKey = HIVE_TEZ_COMMIT_TASK_COUNT + (tableName == null ? "" : "." + tableName);
+              // save info into session conf
+              HiveConf sessionConf = SessionState.get().getConf();
+              sessionConf.set(jobIdKey, jobId);
+              sessionConf.setInt(taskCountKey, status.getProgress().getSucceededTaskCount());
```
Wouldn't this pollute the conf forever (for the session)?

Why not use `queryState` instead?

marton-bod replied:
Good point, I wanted to unset these variables in a finally block during `commitInsertTable`, just forgot to add it.
I did not find a place to put these in `queryState` but maybe I missed something.

pvary replied:
I think we should extend the `queryState` for Iceberg queries. This could be a candidate to put there, and also it could be good to put the snapshots of the table the query is execution on.

What do you think?

marton-bod replied:
Yeah, I'll look into it. One question though: will the queryState be available in the meta hook? Is it available also via the SessionState?

pvary replied:
Maybe we should make it available in a map of queryId/QueryState?
And we can clean that up when we finish a query

marton-bod replied:
I'll look into this in the following PR, once we've replaced the temporary listing solution with the permanent one and upgraded the Tez dependency.

marton-bod commented:
@pvary I've uploaded a set of new commits addressing your review comments + changing the long-term approach with the temporary listing solution, as discussed. The listing solution is mostly based on the current downstream TP solution just simplified and refined a bit to save the information into the session conf. This should make it easy to swap out later with the proper vertex id-based implementation.

marton-bod commented:
This is roughly what the changes would need to look like once we have the new Tez version released:
https://github.com/marton-bod/hive/pull/1

pvary commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:
```diff
@@ -256,4 +265,74 @@ private static PartitionSpec spec(Schema schema, Properties properties,
+
+    // check status to determine whether we need to commit or to abort
+    JobConf jobConf = new JobConf(conf);
+    String queryIdKey = jobConf.get("hive.query.id") + "." + tableName + ".result";
```
Could we use constants here?

pvary replied:
Maybe even "hive.query.id.%.result"? Or "hive.query.id.result."+tableName?

marton-bod replied:
Makes sense. I was thinking of replacing the query id with a constant like `HIVE_TEZ_COMMIT_JOB_RESULT` so it would be `HIVE_TEZ_COMMIT_JOB_ID + "." + tableName`, like for job id and task num to make things consisent.

pvary commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:
```diff
@@ -256,4 +265,74 @@ private static PartitionSpec spec(Schema schema, Properties properties,
+  @Override
+  public void rollbackInsertTable(org.apache.hadoop.hive.metastore.api.Table table, boolean overwrite)
+      throws MetaException {
+    // do nothing
```
Shouldn't we call abortJob here?

marton-bod replied:
I didn't put it there because f there is an execution error we should get a non-0 return code from Tez AM, instead of an exception. But now that I'm thinking about it, I suppose Hive will throw an exception at the end if the code was non-0, but I will test it out to make sure

pvary commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java:
```diff
@@ -105,13 +105,18 @@ public void commitTask(TaskAttemptContext originalContext) throws IOException {
-
-            // Creating the file containing the data files generated by this task for this table
-            createFileForCommit(closedFiles, fileForCommitLocation, table.io());
+            if (table != null) {
```
Could we do this check in `HiveIcebergMetaHook.commitInsertTable`?
Then we would not need any change here...

marton-bod replied:
This happens during task commit, so before the commitInsert hook is called. 

The essential problem here is that `OUTPUT_TABLES` contains all the tables, however, only those tables are serialized into the jobconfig that are relevant for the given task. So it tries to iterate over 1...N tables, but only has access to serialized Table 1 (hence the if).

pvary replied:
Got it, then we should push this change to the Iceberg code as well.

marton-bod replied:
Sure, will do!

marton-bod replied:
Actually, this issue does not occur with mr, only with tez, therefore it currently does not come up in upstream Iceberg. Will fix this once Tez writes have been enabled there too.

pvary replied:
I think this is the same situation that we have been with some other Tez patches. It does not hurt to be there an we keep the two codes from diverging. But it's up to you to decide.

marton-bod replied:
Yeah, I think you're right

pvary commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java:
```diff
@@ -346,25 +372,23 @@ private static ExecutorService tableExecutor(Configuration conf, int maxThreadNu
-  private static Collection<DataFile> dataFiles(ExecutorService executor, String location, JobContext jobContext,
-      FileIO io, boolean throwOnFailure) {
+  private static Collection<DataFile> dataFiles(int numTasks, ExecutorService executor, String location,
+                                                JobContext jobContext, FileIO io, boolean throwOnFailure) {
```
nit: I have seen Iceberg reviewers asking for 4 space padding.

marton-bod replied:
They are a bit confusing about this. The 4 spaces padding is indeed the general rule for line continuation, but I've been asked earlier by Anton on other PRs not to do that for method parameter indentations and do it this way.

pvary replied:
Fun stuff 😄
Whatever!

pvary commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java:
```diff
@@ -239,6 +252,16 @@ public void abortJob(JobContext originalContext, int status) throws IOException
     cleanup(jobContext, jobLocations);
   }
 
+  private Set<FileStatus> listForCommits(JobConf jobConf, String jobLocation) throws IOException {
```
nit: Maybe a javadoc that do not use it for anything else than abort.

marton-bod replied:
Right, good idea
